extension/src/ballistics/bullet/model.rs

```diff
@@ -0,0 +1,216 @@
+                    let grid_height = map.height(grid_cell) as f64;
+                    if grid_height > bullet_position.z() {
+                        let angle = ((grid_height - bullet_position.z()) / 100.0).atan();
+                        wind_attenuation *= angle.cos().abs().powi(2);
+                    }
+                }
+            }
+
+            if height_atl > 0.0 && height_atl < 20.0 {
+                let wind_source_obstacles = bullet_position - wind.normalize() * 25.0;
+                let roughness_length = calculate_roughness_length(
+                    map,
+                    wind_source_obstacles.x(),
+                    wind_source_obstacles.y(),
+                );
+                wind_attenuation *=
+                    0.0f64.max(height_atl / roughness_length.log(20.0 / roughness_length));
+            }
+
+            wind *= wind_attenuation;
+        }
+
+        let mut velocity_offset = Vector3::default();
+
+        {
+            let mut bullet_velocity = self.bullet_velocity_last_frame;
+            let mut time = 0.0;
+            while time < delta_time {
+                let dt = (delta_time - time).min(0.005);
+                let drag_ref = -self.air_friction * bullet_velocity.magnitude_squared();
+                let accel_ref = bullet_velocity.normalize() * drag_ref;
+
+                velocity_offset += accel_ref * dt;
+                bullet_velocity -= accel_ref * dt;
+                bullet_velocity += GRAVITY_ACCEL * dt;
+
+                time += dt;
+            }
+        }
+
+        let mut bullet_velocity = self.bullet_velocity_last_frame;
+        let mut time = 0.0;
+        tof -= delta_time;
+
+        while time < delta_time {
+            let dt = (delta_time - time).min(0.005 * 2.0);
+
+            let mut true_velocity = bullet_velocity - wind;
+
+            let sound_speed = speed_of_sound(temperature);
+            if self.transonic_stability_coefficient < 1.0
+                && true_velocity.magnitude() < 1.2 * sound_speed
+                && true_velocity.magnitude() > sound_speed
+            {
+                let distribution = Uniform::from(-10.0..10.0);
+                let offset = Vector3::new(
+                    distribution.sample(&mut self.rng),
+                    distribution.sample(&mut self.rng),
+                    distribution.sample(&mut self.rng),
+                );
+                let coef = 1.0 - self.transonic_stability_coefficient;
+                let true_speed = true_velocity.magnitude();
+                true_velocity += offset * coef;
+                true_velocity = true_velocity.normalize() * true_speed;
+            }
+
+            let drag = if self.ballistic_coefficients.len() == self.velocity_boundaries.len() + 1 {
+                let mut ballistic_coefficient = self.ballistic_coefficients[0];
+                for (i, boundry) in self.velocity_boundaries.iter().enumerate() {
```

Comment:
[c++ src](https://github.com/acemod/ACE3/blob/master/extensions/advanced_ballistics/AdvancedBallistics.cpp#L577)
```suggestion
                for (i, boundry) in self.velocity_boundaries.iter().enumerate().rev() {
```